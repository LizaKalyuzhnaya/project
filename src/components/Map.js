import React from 'react';
import PropTypes from 'prop-types'
import { compose, withProps, withHandlers } from "recompose"
import { withScriptjs, withGoogleMap, GoogleMap, Marker, InfoWindow, Polyline } from "react-google-maps"

const MyMap = compose(
  withProps({
    googleMapURL: "https://maps.googleapis.com/maps/api/js?key=AIzaSyCnwOD27FbIUsHC-RndOtEYar_ADYBT1Dg",
    loadingElement: <div style={{ height: `80%` }} />,
    containerElement: <div style={{ height: `97vh`, width: `70vw` }} />,
    mapElement: <div style={{ height: `100%` }} />
  }),
  withHandlers(() => {
    const refs = {
      map: undefined,
    }
    const markerRef = [];

    return {
      onMapMounted: () => ref => {
        refs.map = ref
      },
      changeMapCenter: (props) => () => {
        let centerCoord =refs.map.getCenter();
        let [lat, lng] = [centerCoord.lat(), centerCoord.lng()];
        props.updateMapCenter({lat, lng});
      },
      onMarkerMounted: () => (ref) => {
        if (ref) {
          markerRef.push(ref);
        }
      },
      dragMarker: (props) => (id) => {
        let newMarkerPosition = markerRef[id].getPosition();
        let [lat, lng] = [newMarkerPosition.lat(), newMarkerPosition.lng()]
        props.updateMarkerCoordinates(id, {lat, lng});
      },
      onToggleOpen: ( props ) => (id) => {
        props.openInfo(id);
      }
    }
  }),
  withScriptjs,
  withGoogleMap
)((props) => {
  let reducer = (acc, item) => {
    acc.push(item.coordinates);
    return acc;
  }
  let coordinatesForLines = (props.placeNames).reduce(reducer, []);
  return (
      <GoogleMap
        ref={props.onMapMounted}
        defaultZoom={8}
        defaultCenter={{ lat: 59.939095, lng: 30.315868 }}
        onCenterChanged = {props.changeMapCenter}
      >
        {props.placeNames &&  
          props.placeNames.map(item => {
            return(
              <Marker key= {item.id} ref={props.onMarkerMounted} onClick = {() => props.onToggleOpen(item.id)} draggable = {true} onDragEnd ={()=>props.dragMarker(item.id)} position={item.coordinates}>
              {item.isOpen && <InfoWindow onCloseClick = {() => props.onToggleOpen(item.id)} >
                  <div>{item.text}</div>
                </InfoWindow>}
              </Marker> 
              )
          })  
        }
        <Polyline path = {coordinatesForLines} />
      </GoogleMap>
  )
})


MyMap.propTypes = {
  placeNames: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      text: PropTypes.string.isRequired,
      isOpen: PropTypes.bool.isRequired,
      coordinates: PropTypes.object.isRequired
    }).isRequired
  ).isRequired,
  updateMarkerCoordinates: PropTypes.func.isRequired,
  updateMapCenter: PropTypes.func.isRequired,
  openInfo: PropTypes.func.isRequired
}


export default MyMap;

