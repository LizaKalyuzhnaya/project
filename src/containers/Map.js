import { connect } from 'react-redux'
import myMap from '../components/Map'
import {updateMapCenter} from '../actions'
import {updateMarkerCoordinates} from '../actions'
import {openInfo} from '../actions'

const mapStateToProps = state => ({
    placeNames: state.placeNames
})
const mapDispatchToProps = dispatch => ({
    updateMapCenter: (coordinates) => dispatch(updateMapCenter(coordinates)),
    updateMarkerCoordinates: (id, coordinates) => dispatch(updateMarkerCoordinates(id, coordinates)),
    openInfo: (id) => dispatch(openInfo(id))
  })

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(myMap)