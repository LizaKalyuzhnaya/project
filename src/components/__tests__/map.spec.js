import React from 'react'
import { shallow } from 'enzyme'
import Map from '../Map'

describe('Map', () => {
    describe('Map when balloon closed', () => {
        const props = {
            placeNames: [{
                id: 0,
                text: 'A',
                coordinates: { lat: 20, lng: 30 },
                isOpen: false
            }]        
        }

        const myComponent = shallow(<Map {...props} />)

        it('renders properly', () => {
            expect(myComponent).toMatchSnapshot()
        })
    })

    describe('Map when balloon open', () => {
        const props = {
            placeNames: [{
                id: 0,
                text: 'A',
                coordinates: { lat: 20, lng: 30 },
                isOpen: true
            }]        
        }

        const myComponent = shallow(<Map {...props} />)

        it('renders properly', () => {
            expect(myComponent).toMatchSnapshot()
        })
    })
})