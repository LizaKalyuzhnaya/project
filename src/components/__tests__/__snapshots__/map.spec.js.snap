// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Map when balloon closed renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withProps(withHandlers(withScriptjs(withGoogleMap(Component))))
    placeNames={
      Array [
        Object {
          "coordinates": Object {
            "lat": 20,
            "lng": 30,
          },
          "id": 0,
          "isOpen": false,
          "text": "A",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "containerElement": <div
        style={
          Object {
            "height": "97vh",
            "width": "70vw",
          }
        }
      />,
      "googleMapURL": "https://maps.googleapis.com/maps/api/js?key=AIzaSyCnwOD27FbIUsHC-RndOtEYar_ADYBT1Dg",
      "loadingElement": <div
        style={
          Object {
            "height": "80%",
          }
        }
      />,
      "mapElement": <div
        style={
          Object {
            "height": "100%",
          }
        }
      />,
      "placeNames": Array [
        Object {
          "coordinates": Object {
            "lat": 20,
            "lng": 30,
          },
          "id": 0,
          "isOpen": false,
          "text": "A",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "containerElement": <div
          style={
            Object {
              "height": "97vh",
              "width": "70vw",
            }
          }
        />,
        "googleMapURL": "https://maps.googleapis.com/maps/api/js?key=AIzaSyCnwOD27FbIUsHC-RndOtEYar_ADYBT1Dg",
        "loadingElement": <div
          style={
            Object {
              "height": "80%",
            }
          }
        />,
        "mapElement": <div
          style={
            Object {
              "height": "100%",
            }
          }
        />,
        "placeNames": Array [
          Object {
            "coordinates": Object {
              "lat": 20,
              "lng": 30,
            },
            "id": 0,
            "isOpen": false,
            "text": "A",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Map Map when balloon open renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withProps(withHandlers(withScriptjs(withGoogleMap(Component))))
    placeNames={
      Array [
        Object {
          "coordinates": Object {
            "lat": 20,
            "lng": 30,
          },
          "id": 0,
          "isOpen": true,
          "text": "A",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "containerElement": <div
        style={
          Object {
            "height": "97vh",
            "width": "70vw",
          }
        }
      />,
      "googleMapURL": "https://maps.googleapis.com/maps/api/js?key=AIzaSyCnwOD27FbIUsHC-RndOtEYar_ADYBT1Dg",
      "loadingElement": <div
        style={
          Object {
            "height": "80%",
          }
        }
      />,
      "mapElement": <div
        style={
          Object {
            "height": "100%",
          }
        }
      />,
      "placeNames": Array [
        Object {
          "coordinates": Object {
            "lat": 20,
            "lng": 30,
          },
          "id": 0,
          "isOpen": true,
          "text": "A",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "containerElement": <div
          style={
            Object {
              "height": "97vh",
              "width": "70vw",
            }
          }
        />,
        "googleMapURL": "https://maps.googleapis.com/maps/api/js?key=AIzaSyCnwOD27FbIUsHC-RndOtEYar_ADYBT1Dg",
        "loadingElement": <div
          style={
            Object {
              "height": "80%",
            }
          }
        />,
        "mapElement": <div
          style={
            Object {
              "height": "100%",
            }
          }
        />,
        "placeNames": Array [
          Object {
            "coordinates": Object {
              "lat": 20,
              "lng": 30,
            },
            "id": 0,
            "isOpen": true,
            "text": "A",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
